### ISSUE TITLE
Help with making a static visualisation of topics

### ISSUE THREAD

**Issue by schetudiante** [MaartenGr/BERTopic#126]
Dear Maarten
Firstly congratulations on this nice package it is packing our research easy!

I am a student researcher in NLP trying to use your package for clustering sentences., 
I am following the [code from your original tutorial](https://colab.research.google.com/drive/1FieRA9fLdkQEGDIMYl0I3MCjSUKVF8C-?usp=sharing/)to make clusters from my sentences but I need help to visualise the most frequent topics that [may look like this](https://raw.githubusercontent.com/UKPLab/sentence-transformers/master/docs/img/20news_top2vec.png), can you please share some reference code to help me to make such a visualization?
I need it for my research and it would be great if you could help.

Thank you

**Comment by MaartenGr** [comment]
Thank you for the kind words!

This is not something that is supported in BERTopic as all documents will need to be saved in the model. If many millions of documents are saved, then that could explode the size of the model. 

Having said that, to do this, you can follow along with the following code:

```python
import numpy as np
import pandas as pd
from umap import UMAP
from bertopic import BERTopic
import matplotlib.pyplot as plt

%matplotlib inline

# First, we need to extract topics
topic_model = BERTopic(verbose=True, embedding_model="stsb-mpnet-base-v2")
topics, _ = topic_model.fit_transform(docs)

# Then, we extract the embeddings for each document
embeddings = topic_model._extract_embeddings(docs, method="document")

# Reducing dimensionality to 2d -> Play around with n_neighbors and min_dist
# Note that these 2D embeddings are very sensitive to hyperparameters
umap_embeddings = UMAP(n_neighbors=15, n_components=2, min_dist=0.0, metric='cosine').fit_transform(embeddings)

# Combine data
df = pd.DataFrame(umap_embeddings, columns=["x", "y"])
df["topic"] = topics

# Visualize topics
fig, ax = plt.subplots(figsize=(10, 10))
ax.scatter(df['x'], df['y'], c=df['topic'], s=1, alpha=.3, cmap="Set1")
plt.show()
```

To show only the most frequent topics, I would suggest replacing low frequent topics in `topics` with -1. Then, make sure that -1 gets a greyish color such that the topics will be more pronounced.

**Comment by schetudiante** [comment]
wow! Thank you So much Maarten :)

I ran the code above with my sentences ( a smallish dataset of about 23,729) , I played around with neighbors n+ from 15 to 20 and min_dist = 0.0 - 0.8 
but uptill now no figure I was able to see a label (almost all of them look like this)
 
![download](https://user-images.githubusercontent.com/82269015/119523041-bc704200-bd7c-11eb-8e90-aff71b7bbdf1.png)

**Comment by MaartenGr** [comment]
No problem :) That is to be expected since the code I shared does not add the labels to specific clusters. This is something you will have to do in matplotlib yourself. You can use `ax.legend()` to add the legend. If you want the labels to hover over the figure, then you will have to do that either manually or see if you can position them using the centroid of each cluster.

**Comment by schetudiante** [comment]
Thank you so much! Maarten
Best of luck

**Comment by schetudiante** [comment]
Hey Maarten
Could you help me with making this diagram: 
![sidekick_sHFRIxtkH8](https://user-images.githubusercontent.com/82269015/119725493-4d741580-be70-11eb-83b9-cb51f7cf615c.png)


 only contain the top 10 or 20 (frequent) occuring topic clusters and maybe color them different?
It is for my research .
Thank you

**Comment by MaartenGr** [comment]
To get such a visualization I would advise you to install the upcoming version of BERTopic (v0.8 #120). This version allows you to select and visualize the top X frequent topics with `visualize_topics(top_n_topics=10)`. 

Then, you will need to dive into the [code](https://github.com/MaartenGr/BERTopic/blob/v0.8/bertopic/plotting/_topics.py) yourself to adapt the colors. There, you need to remove setting the color to the greyish-blue it is now:
https://github.com/MaartenGr/BERTopic/blob/72a5773c1921e8d692f75822e570712738eca303/bertopic/plotting/_topics.py#L89

**Comment by schetudiante** [comment]
Thanks Maarten for your graceful reply

But can you tell me how I can visualise topics outside jupyter (notebook) and save them in a normal script.
and maybe have legends the words?

thanks

**Comment by MaartenGr** [comment]
You can render the visualization using multiple methods described [here](https://plotly.com/python/renderers/). To save a figure simply call `fig.write_html("path/to/file.html")` which saves an interactive version of the figure. If you want a static version take a look [here](https://plotly.com/python/static-image-export/).

The legend can be a bit more tricky since it requires separately adding the traces as shown [here](https://plotly.com/python/legend/).

**Comment by schetudiante** [comment]
Thank you so much Maarten.

**Comment by MaartenGr** [comment]
@schetudiante 

I had some time to play around with creating the visualization that you were looking for. Let me know how it works out! 

```python
import numpy as np
import pandas as pd
from umap import UMAP
from bertopic import BERTopic
from sklearn.datasets import fetch_20newsgroups

import matplotlib
import matplotlib.pyplot as plt

%matplotlib inline

# Create topic model
docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))["data"]
topic_model = BERTopic(embedding_model="paraphrase-MiniLM-L6-v2", verbose=True)
topics, _ = topic_model.fit_transform(docs)

# Prepare data for plotting
embeddings = topic_model._extract_embeddings(docs, method="document")
umap_model = UMAP(n_neighbors=10, n_components=2, min_dist=0.0, metric='cosine').fit(embeddings)
df = pd.DataFrame(umap_model.embedding_, columns=["x", "y"])
df["topic"] = topics

# Plot parameters
top_n = 10
fontsize = 12

# Slice data
to_plot = df.copy()
to_plot[df.topic >= top_n] = -1
outliers = to_plot.loc[to_plot.topic == -1]
non_outliers = to_plot.loc[to_plot.topic != -1]

# Visualize topics
cmap = matplotlib.colors.ListedColormap(['#FF5722', # Red
                                         '#03A9F4', # Blue
                                         '#4CAF50', # Green
                                         '#80CBC4', # FFEB3B
                                         '#673AB7', # Purple
                                         '#795548', # Brown
                                         '#E91E63', # Pink
                                         '#212121', # Black
                                         '#00BCD4', # Light Blue
                                         '#CDDC39', # Yellow/Red
                                         '#AED581', # Light Green
                                         '#FFE082', # Light Orange
                                         '#BCAAA4', # Light Brown
                                         '#B39DDB', # Light Purple
                                         '#F48FB1', # Light Pink
                                         ])

# Visualize outliers + inliers
fig, ax = plt.subplots(figsize=(15, 15))
scatter_outliers = ax.scatter(outliers['x'], outliers['y'], c="#E0E0E0", s=1, alpha=.3)
scatter = ax.scatter(non_outliers['x'], non_outliers['y'], c=non_outliers['topic'], s=1, alpha=.3, cmap=cmap)

# Add topic names to clusters
centroids = to_plot.groupby("topic").mean().reset_index().iloc[1:]
for row in centroids.iterrows():
    topic = int(row[1].topic)
    text = f"{topic}: " + "_".join([x[0] for x in topic_model.get_topic(topic)[:3]])
    ax.text(row[1].x, row[1].y*1.01, text, fontsize=fontsize, horizontalalignment='center')

ax.text(0.99, 0.01, f"BERTopic - Top {top_n} topics", transform=ax.transAxes, horizontalalignment="right", color="black")
plt.xticks([], [])
plt.yticks([], [])
plt.show()
```

![bertopic_viz](https://user-images.githubusercontent.com/25746895/120966149-1baf5880-c766-11eb-9d01-0898deb24adb.png)

**Comment by schetudiante** [comment]
I don't know what to say but..
Thanks a TON Sir!
<3

**Comment by MaartenGr** [comment]
No problem, just glad I could be of some help!

**Comment by randomrdrand** [comment]
> ```python
> # Prepare data for plotting
> embeddings = topic_model._extract_embeddings(docs, method="document")
> umap_model = UMAP(n_neighbors=10, n_components=2, min_dist=0.0, metric='cosine').fit(embeddings)
> df = pd.DataFrame(umap_model.embedding_, columns=["x", "y"])
> df["topic"] = topics
> 
> # Plot parameters
> top_n = 10
> fontsize = 12
> 
> # Slice data
> to_plot = df.copy()
> to_plot[df.topic >= top_n] = -1
> outliers = to_plot.loc[to_plot.topic == -1]
> non_outliers = to_plot.loc[to_plot.topic != -1]
> 
> # Visualize topics
> cmap = matplotlib.colors.ListedColormap(['#FF5722', # Red
>                                          '#03A9F4', # Blue
>                                          '#4CAF50', # Green
>                                          '#80CBC4', # FFEB3B
>                                          '#673AB7', # Purple
>                                          '#795548', # Brown
>                                          '#E91E63', # Pink
>                                          '#212121', # Black
>                                          '#00BCD4', # Light Blue
>                                          '#CDDC39', # Yellow/Red
>                                          '#AED581', # Light Green
>                                          '#FFE082', # Light Orange
>                                          '#BCAAA4', # Light Brown
>                                          '#B39DDB', # Light Purple
>                                          '#F48FB1', # Light Pink
>                                          ])
> 
> # Visualize outliers + inliers
> fig, ax = plt.subplots(figsize=(15, 15))
> scatter_outliers = ax.scatter(outliers['x'], outliers['y'], c="#E0E0E0", s=1, alpha=.3)
> scatter = ax.scatter(non_outliers['x'], non_outliers['y'], c=non_outliers['topic'], s=1, alpha=.3, cmap=cmap)
> 
> # Add topic names to clusters
> centroids = to_plot.groupby("topic").mean().reset_index().iloc[1:]
> for row in centroids.iterrows():
>     topic = int(row[1].topic)
>     text = f"{topic}: " + "_".join([x[0] for x in topic_model.get_topic(topic)[:3]])
>     ax.text(row[1].x, row[1].y*1.01, text, fontsize=fontsize, horizontalalignment='center')
> 
> ax.text(0.99, 0.01, f"BERTopic - Top {top_n} topics", transform=ax.transAxes, horizontalalignment="right", color="black")
> plt.xticks([], [])
> plt.yticks([], [])
> plt.show()
> ```

I am just wondering, the dots are topics right? or they are documents?

**Comment by MaartenGr** [comment]
They dots in the figure are documents and the colors represent the topics.

**Comment by jeremycah1ll5** [comment]
Is it possible to plot the hierarchical plot using Matplotlib? Do you have any suggestions to plotting the data when not using a Jupyter notebook?

**Comment by MaartenGr** [comment]
@jeremycah1ll5 You can do [static exports with Plotly](https://plotly.com/python/static-image-export/). Other than that, it would require some customization to get it exported to matplotlib. Checking out the source code might help.
